### ISSUE TITLE
Variables and columns are not actually permuted

### ISSUE THREAD

**Issue by dchetelat** [ds4dm/ecole#159]
### Problem
By default, Ecole changes the SCIP parameters to randomize the ordering of variables and columns. Namely, it calls
`EnvironmentDynamics.set_dynamics_random_state` in the python code,
https://github.com/ds4dm/ecole/blob/b13c6fba2a3e2b3ccb4ecfb3d1b941adfcbdd363/python/src/ecole/environment.py#L103-L109
where `set_dynamics_random_state` is defined as follows:
https://github.com/ds4dm/ecole/blob/2201252db3427786b437174e2a0541aa47e29bc7/libecole/include/ecole/dynamics/dynamics.hpp#L32-L40

The problem is that this doesn't do anything! Because setting `randomization/permutevars`/`randomization/permuteconss` must happen before the instance is loaded. 

### Code example
A simple code example in `pyscipopt` is as follows (thanks @lascavana ):

```python
import pyscipopt as scip
import numpy as np
import random

class CustomBranchRule(scip.Branchrule):
    def branchexeclp(self, allowaddcons):
        candidate_vars, *_ = self.model.getLPBranchCands()
        for var in candidate_vars:
            var_num = int(str(var).split('x')[1])
            var_pos = var.getCol().getLPPos()
            print(f"Variable {var} is at position {var_pos} (diff {var_num-var_pos})")

        self.model.interruptSolve()
        branch_var = candidate_vars[0]
        self.model.branchVar(branch_var)
        result = scip.SCIP_RESULT.BRANCHED
        return {"result": result}



filename = 'indset.lp'
rng = np.random.RandomState(0)

model = scip.Model()
model.setParam('display/verblevel', 0)
model.setParam('separating/maxcuts', 0)
model.setParam('separating/maxcutsroot', 0)
model.setParam('presolving/maxrestarts', 0)

# Randomization #
model.setParam('randomization/permuteconss', True)
model.setParam('randomization/permutevars', True)
model.setParam('randomization/randomseedshift', rng.randint(0, 2**16))
model.setParam('randomization/permutationseed', rng.randint(0, 2**16))
model.setParam('randomization/lpseed', rng.randint(0, 2**16))

model.readProblem(filename)
branchrule = CustomBranchRule()
model.includeBranchrule(branchrule=branchrule,
                        name="Custom",     # name of the branching rule
                        desc="",           # description of the branching rule
                        priority=100000,   # priority: set to this to make it default
                        maxdepth=-1,       # maximum depth up to which it will be used, or -1 for no restriction
                        maxbounddist=1)    # maximal relative distance from current node's dual bound to primal
                                           # bound compared to best node's dual bound for applying branching rule
                                           # (0.0: only on current best node, 1.0: on all nodes)
model.optimize()
model.freeProb()
```
which outputs
```
Variable t_x192 is at position 0 (diff 192)
Variable t_x33 is at position 1 (diff 32)
Variable t_x113 is at position 2 (diff 111)
Variable t_x226 is at position 3 (diff 223)
Variable t_x183 is at position 4 (diff 179)
Variable t_x154 is at position 5 (diff 149)
Variable t_x68 is at position 6 (diff 62)
Variable t_x32 is at position 7 (diff 25)
Variable t_x107 is at position 8 (diff 99)
Variable t_x150 is at position 9 (diff 141)
Variable t_x196 is at position 10 (diff 186)
Variable t_x229 is at position 11 (diff 218)
Variable t_x9 is at position 12 (diff -3)
Variable t_x41 is at position 13 (diff 28)
Variable t_x71 is at position 14 (diff 57)
...
```
whereas if you set the randomization parameters after the loading, 
```python
import pyscipopt as scip
import numpy as np
import random

class CustomBranchRule(scip.Branchrule):
    def branchexeclp(self, allowaddcons):
        candidate_vars, *_ = self.model.getLPBranchCands()
        for var in candidate_vars:
            var_num = int(str(var).split('x')[1])
            var_pos = var.getCol().getLPPos()
            print(f"Variable {var} is at position {var_pos} (diff {var_num-var_pos})")

        self.model.interruptSolve()
        branch_var = candidate_vars[0]
        self.model.branchVar(branch_var)
        result = scip.SCIP_RESULT.BRANCHED
        return {"result": result}



filename = 'indset.lp'
rng = np.random.RandomState(0)

model = scip.Model()
model.setParam('display/verblevel', 0)
model.setParam('separating/maxcuts', 0)
model.setParam('separating/maxcutsroot', 0)
model.setParam('presolving/maxrestarts', 0)

model.readProblem(filename)

# Randomization #
model.setParam('randomization/permuteconss', True)
model.setParam('randomization/permutevars', True)
model.setParam('randomization/randomseedshift', rng.randint(0, 2**16))
model.setParam('randomization/permutationseed', rng.randint(0, 2**16))
model.setParam('randomization/lpseed', rng.randint(0, 2**16))

branchrule = CustomBranchRule()
model.includeBranchrule(branchrule=branchrule,
                        name="Custom",     # name of the branching rule
                        desc="",           # description of the branching rule
                        priority=100000,   # priority: set to this to make it default
                        maxdepth=-1,       # maximum depth up to which it will be used, or -1 for no restriction
                        maxbounddist=1)    # maximal relative distance from current node's dual bound to primal
                                           # bound compared to best node's dual bound for applying branching rule
                                           # (0.0: only on current best node, 1.0: on all nodes)
model.optimize()
model.freeProb()
```
you get
```
Variable t_x1 is at position 0 (diff 1)
Variable t_x2 is at position 1 (diff 1)
Variable t_x3 is at position 2 (diff 1)
Variable t_x4 is at position 3 (diff 1)
Variable t_x5 is at position 4 (diff 1)
Variable t_x6 is at position 5 (diff 1)
Variable t_x7 is at position 6 (diff 1)
Variable t_x8 is at position 7 (diff 1)
Variable t_x9 is at position 8 (diff 1)
Variable t_x10 is at position 9 (diff 1)
Variable t_x11 is at position 10 (diff 1)
Variable t_x12 is at position 11 (diff 1)
Variable t_x13 is at position 12 (diff 1)
Variable t_x14 is at position 13 (diff 1)
Variable t_x15 is at position 14 (diff 1)
Variable t_x16 is at position 15 (diff 1)
Variable t_x17 is at position 16 (diff 1)
Variable t_x18 is at position 17 (diff 1)
Variable t_x19 is at position 18 (diff 1)
Variable t_x20 is at position 19 (diff 1)
Variable t_x21 is at position 20 (diff 1)
...
```

### Extension

More generally, there might be other SCIP parameters that might need to be called after the model is created, but before the instance is loaded. (Although I don't have any examples). I don't know if the reverse is true. One possibility would be to refactor the code so that both manual and `set_dynamics_random_state` SCIP parameters get set always before the instance is loaded.

**Comment by AntoinePrv** [comment]
So far it was set like this because for branching, these permutation where done during transform (before presolve).

The problem is that "_loading_" is a concept unknown to the environment, which takes as input a `scip.Model` that could come from an instance generator (filename are simply an alias for [`from_file`](https://github.com/ds4dm/ecole/blob/2201252db3427786b437174e2a0541aa47e29bc7/python/src/ecole/environment.py#L106).

**Some ideas**:
 - Call [`SCIPtransformProb`](https://www.scipopt.org/doc-7.0.2/html/group__PublicSolveMethods.php#ga488d46edda94eb64ed8c5e272f7d999b) in `Configuring.reset`;
 - Have a (file) instance loader, similar to generator, that load and randomize instances. Instance generators have their own source of randomness, so it would make sense to put the "loading" randomness there as well.

**Comment by dchetelat** [comment]
> So far it was set like this because for branching, these permutation where done during transform (before presolve).
I'm not sure I understand - how could this have worked for branching, since if you call it at any point after `SCIPreadProblem`, it doesn't do anything?

Maybe I wasn't very clear right now but it doesn't work.
```python
import ecole
import numpy as np

class CandidatesObs:
    def before_reset(self, model):
        pass

    def extract(self, model, done):
        if done:
            return None
        scip = model.as_pyscipopt()
        candidate_vars, *_ = scip.getLPBranchCands()
        for var in candidate_vars:
            var_num = int(str(var).split('x')[1])
            var_pos = var.getCol().getLPPos()
            print(f"Variable {var} is at position {var_pos} (diff {var_num-var_pos})")


env = ecole.environment.Branching(
    scip_params={
        'separating/maxcuts': 0,
        'separating/maxcutsroot': 0,
        'presolving/maxrestarts': 0,
        'limits/time': 3600,
        'timing/clocktype': 2},
    pseudo_candidates=False,
    observation_function=CandidatesObs()
)

instance = 'indset.lp'
rng = np.random.RandomState(seed=0)
seed = rng.randint(0, 2**16)
print(f'Seed: {seed}')
env.seed(seed)
observation, action_set, _, done, info = env.reset(instance)
```
yields
```
Seed: 2732
Variable t_x1 is at position 0 (diff 1)
Variable t_x2 is at position 1 (diff 1)
Variable t_x3 is at position 2 (diff 1)
Variable t_x4 is at position 3 (diff 1)
Variable t_x5 is at position 4 (diff 1)
Variable t_x6 is at position 5 (diff 1)
Variable t_x7 is at position 6 (diff 1)
Variable t_x8 is at position 7 (diff 1)
Variable t_x9 is at position 8 (diff 1)
Variable t_x10 is at position 9 (diff 1)
Variable t_x11 is at position 10 (diff 1)
Variable t_x12 is at position 11 (diff 1)
...
```

**Comment by AntoinePrv** [comment]
Sorry, I thought this arises when using `Configuring` style where not SCIP function are called.

What about idea 2 then:
> Have a (file) instance loader, similar to generator, that load and randomize instances. Instance generators have their own source of randomness, so it would make sense to put the "loading" randomness there as well.

**Comment by gasse** [comment]
> ### Problem
> By default, Ecole changes the SCIP parameters to randomize the ordering of variables and columns. Namely, it calls
> `EnvironmentDynamics.set_dynamics_random_state` in the python code,
> https://github.com/ds4dm/ecole/blob/b13c6fba2a3e2b3ccb4ecfb3d1b941adfcbdd363/python/src/ecole/environment.py#L103-L109
> 
> 
> where `set_dynamics_random_state` is defined as follows:
> https://github.com/ds4dm/ecole/blob/2201252db3427786b437174e2a0541aa47e29bc7/libecole/include/ecole/dynamics/dynamics.hpp#L32-L40

What about treatingh those two cases differently ? If one loads an instance file, first set the SCIP parameters, then load the instance. If one uses an existing SCIP instance, copy that instance, then set the parameters. In the second case, we assume that row/column permutation is the responsibility of the user (or the instance generator). This is also compatible with @AntoinePrv 's second suggestion, which is to have an instance generator that does exactly that: load MILP instances + handle a random permutation.

**Comment by gasse** [comment]
> Have a (file) instance loader, similar to generator, that load and randomize instances. Instance generators have their own source of randomness, so it would make sense to put the "loading" randomness there as well.

I like that idea, however I think people should also be able to load an instance into Ecole from a file, with permutation, without having to go through a generator. Simply for debugging purposes for example. Then the "file reader + randomization" generator would be very simple, as it would simply reuse that feature.

**Comment by dchetelat** [comment]
Fixed with SCIP 7.0.3.
